packages/react-aria-components/src/Tree.tsx

```diff
@@ -273,7 +291,114 @@ function TreeInner<T extends object>({props, collection, treeRef: ref}: TreeInne
         dropTargetDelegate: {
           getDropTargetFromPoint(x, y, isValidDropTarget) {
             let target = dropTargetDelegate.getDropTargetFromPoint(x, y, isValidDropTarget);
+            let tracking = pointerTrackingRef.current;
+            
+            // Calculate movement direction
+            let deltaY = y - tracking.lastY;
+            let currentMovement: 'up' | 'down' | null = null;
+            
+            if (Math.abs(deltaY) > 3) {
+              currentMovement = deltaY > 0 ? 'down' : 'up';
+              tracking.movementDirection = currentMovement;
+            }
+            
+            // Handle ambiguous drop position: 'after last child' or 'after parent'
             if (target?.type === 'item' && target.dropPosition === 'after') {
+              let currentItem = state.collection.getItem(target.key);
+
+              // If dropping "after" an expanded item with children,
+              // change target to be "before" its first actual child item.
+              if (currentItem && currentItem.hasChildNodes && state.expandedKeys.has(currentItem.key) && state.collection.getChildren) {
+                let firstChildItemNode: Node<any> | null = null;
+                for (let child of state.collection.getChildren(currentItem.key)) {
+                  if (child.type === 'item' || child.type === 'loader') {
+                    firstChildItemNode = child;
+                    break;
+                  }
+                }
+
+                if (firstChildItemNode) {
+                  const newTarget = {
+                    type: 'item',
+                    key: firstChildItemNode.key,
+                    dropPosition: 'before'
+                  } as const;
+
+                  if (isValidDropTarget(newTarget)) {
+                    tracking.boundaryContext = null;
+                    tracking.lastY = y;
+                    return newTarget;
+                  }
+                }
+              }
+
+              let parentKey = currentItem?.parentKey;
+
+              if (parentKey) {
+                let parentItem = state.collection.getItem(parentKey);
+                let isParentExpanded = parentItem && state.expandedKeys.has(parentKey);
+                
+                if (isParentExpanded) {
+                  let nextKey = state.collection.getKeyAfter(target.key);
+                  let nextItem = nextKey ? state.collection.getItem(nextKey) : null;
+                  let isLastChild = !nextItem || nextItem.parentKey !== parentKey;
+                  
+                  if (isLastChild) {                    
+                    let afterParentTarget = {
+                      type: 'item',
+                      key: parentKey,
+                      dropPosition: 'after'
+                    } as const;
+                    
+                    // eslint-disable-next-line max-depth
```

Comment:
can we split out any helper functions? might help with readability as well to encapsulate some of this logic. It doesn't have to be shared code in order to split it out

Reply:
Should be cleaner in TreeDropTargetDelegate.ts now